File: RPi/lex/js/lex.js
// lex.js 
var AWS = require('aws-sdk'),
  fs = require('fs'),
  ts = require('tailstream'),
  exec = require('child_process').exec;

var FULFILLED = 'Fulfilled',
  RESPONSE_FILE = 'response.mpeg',
  REMOVE_REQUEST_FILE = 'rm request.wav',
  SOX_COMMAND = 'sox -d -t wavpcm -c 1 -b 16 -r 16000 -e signed-integer --endian little - silence 1 0 12% 5 0.3t 15% > request.wav',
  streaming = false,
  inputStream,
  lexruntime = new AWS.LexRuntime({
    region: 'eu-west-1',
    credentials: new AWS.Credentials('AKIAIU2T23GOY6DWSBEQ','GHd4lR2XlCNX/tTA871NifN7ibYPSuoEIdd3BINC' , null)
  });

var setupStream = function() {
  streaming = true;
  inputStream = ts.createReadStream('./request.wav');
  var params = {
    botAlias: '$LATEST',
    botName: 'Libby' ,
    userId: 'lexHeadTesting',
    contentType: 'audio/l16; rate=16000; channels=1',
    inputStream: inputStream
  };

  lexruntime.postContent(params, function(err, data) {
    //var res = exec('sox ./request.wav -d');
    //console.log("response: " + JSON.stringify(res));
    console.log(data);
    if (err) {
      console.log(err, err.stack);
      process.exit(1);
    } else {
      fs.writeFile(RESPONSE_FILE, data.audioStream, function(err) {
        if (err) {
          return console.log(err);
          process.exit(1);
        }
      });
      var playback = exec('sudo mpg321 ' + RESPONSE_FILE);
      playback.on('close', function(code) {
        exec('rm ' + RESPONSE_FILE);
        if (data.dialogState !== FULFILLED) {
          streaming = false;
          //record();
        }
      });
    }
  });
}

var record = function() {
  var recording = exec(SOX_COMMAND);
  recording.stderr.on('data', function(data) {
    console.log("jee" + data);
    if (!streaming) {
      setupStream();
    }
  });
  recording.on('close', function(code) {
    console.log("closing recording\n\n\n");
    inputStream.done();
    exec(REMOVE_REQUEST_FILE);
  });
}
record();

File: rpi/lex/python/lex.py
import boto3
from subprocess import run, PIPE, CompletedProcess
import wave


class Lex():
    '''
    This class can be used to send voice data to AWS Lex.
    The credentials for aws are loaded from ~/.aws/credentials
    and region from ~/.aws/region
    '''
    def __init__(self):
        self.client = boto3.client('lex-runtime')
        self.bot_name = 'Libby'
        self.bot_alias = 'Dev'
        self.user_id = 'lexpi'
        self.content_type = 'audio/l16; rate=16000; channels=1'
        self.input_stream = b''
        self.input_fname = 'request.wav'
        self.response_fname = 'response.wav'
        self.sox_command = 'sox -d -t wavpcm -c 1 -b 16 -r 16000 -e signed-integer --endian little - silence 1 0 6% 5 0.275t 7%'

    '''
    Create a stream while recording and return it
    '''
    def record(self):
        return run(self.sox_command.split(), stdout=PIPE).stdout

    def post_content(self):
        data = self.record()
        response = self.client.post_content(
            botName=self.bot_name,
            botAlias=self.bot_alias,
            userId=self.user_id,
            contentType=self.content_type,
            inputStream=data
        )

        print(response)
        return response

    '''
    Play back the response's audio stream
    '''
    def play_response(self, response):
        audio_stream = response['audioStream'].read()
        response['audioStream'].close()

        f = wave.open(self.response_fname, 'wb')
        f.setnchannels(2)
        f.setsampwidth(2)
        f.setframerate(16000)
        f.setnframes(0)

        f.writeframesraw(audio_stream)
        f.close()
        run(['mpg321', self.response_fname])

        state = response['dialogState']
        return state



